rfcs/0165-Anonymous-scopes.md

```diff
@@ -0,0 +1,48 @@
+# RFC 165 - Anonymous scopes
+* Comments: [#165](https://github.com/taskcluster/taskcluster-rfcs/pull/165)
+* Proposed by: @ricky26
+
+# Summary
+
+We currently assign no scopes to calls with no credentials. With this proposal
+we would assign a single role to all calls and any scopes expanded from that.
+Additionally, we will move all API calls behind at least one scope.
+
+## Motivation
+
+This is primarily aimed at private TaskCluster installations. With this, it will
+be possible to prevent any useful access to the TaskCluster installation without
+credentials.
+
+# Details
+## Implementation
+A single scope will be assumed for all calls (with or without credentials):
+`assume:anonymous`. A new API call will be added to the auth service
+(`authenticateAnonymous`) which will return `{"scopes":...,"expiry":...}`.
+The authentication middleware in each service will call `authenticateAnonymous` but cache the
+result until `expiry`. A deploy-time variable will be added (`auth.scopes_ttl`)
+which configures how long the auth service declares looked up scopes to be
+valid. 
```

Comment:
Yeah, that makes sense. The only reason I didn't look to repurpose the existing authentication endpoint is that it's explicitly called `authenticateHawk`, which I figured might cause more confusion down the line.

`service:<service-name>:<api-name>` sounds good. I was worrying about the number of scopes we'd be adding but I guess that doesn't actually have a cost, particularly if they're just granted with a wildcard.

(Thanks for putting links to the code in!)

Reply:
This is particularly sensible since there are cases (`queue.getArtifact` for example) where a method can be called with or without credentials.  Having a single way that those cases are handled will improve understand-ability.  Then caching of such requests is just an implementation detail.  I like the idea of the `authenticateHawk` response containing information on how long the response can be cached (but note that this is not the same as the existing `expires` property, which might be 1000 years later).

As for the scopes for particular API methods.. historically it's been really hard to change scopes after they are deployed.  I think I'd prefer to make an attempt to get the "right" scope expression for each currently-open-access API method.  In many cases there are obvious options that parallel the existing modification scopes, such as `worker-manager:get-worker-pool:<workerPoolId>`.  I do *not* think that those scopes need to be worked out in this RFC - that can happen during the implementation phase, with a preference for deferring any tricky decisions to followup work.

Once that's done, we can just modify tc-lib-api to assert that there's a nonempty scope expression for every API method (except `authenticateHawk` itself).

Reply:
> Yeah, that makes sense. The only reason I didn't look to repurpose the existing authentication endpoint is that it's explicitly called `authenticateHawk`, which I figured might cause more confusion down the line.

That's a very fair point.

Reply:
Also, my suggestion of having the signature validator called by `authenticate` when there is no signature, is probably also a bit awry.

So I tend to prefer your idea of a separate endpoint for retrieving the expanded `assume:anonymous` scopes, fetching/caching that early on in [`remoteAuthentication.authenticate`](https://github.com/taskcluster/taskcluster/blob/5a25a717299e9eae61d7ed0935ceb028b8319a26/libraries/api/src/middleware/auth.js#L114-L322) and aggregating the scopes to any other scopes that method would have returned already.

This does mean anything calling `auth.authenticateHawk` won't see scopes associated to the anonymous role, but that probably makes sense, since those scopes are not provided by virtue of some hawk authentication, but come "for free" even without being authenticated in hawk, and can be found via the new API endpoint you suggest.

I think that might also solve djmitche's concern, since `remoteAuthentication.authenticate` will consider aggregated anonymous and hawk scopes when considering scope satisfaction in API calls, even when explicit AuthorizedScopes do not include the anonymous scopes.

Reply:
My expectation was that we would just add `assume:anonymous` inside the `authenticateHawk` call, adding it to the returned list of scopes in the token, before expansion. I figured that the caching is much less relevant there: the additional role expansion should be fairly insignificant compared to the overhead of the authentication call.

Mixing in the anonymous scopes in the middleware is pretty straight-forward too though...